### PR TITLE
add 5 TP algorithm

### PR DESCRIPTION
**Pull request by zhoney**
1. same neighbors
2. jaccard similarity
3. all shortest paths
4. weighted shortest path
5. single source shortest paths

Change-Id: I13a04267c6c91cc8bb00ca2aed1ef54c8e6eeeb6

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/JaccardSimilarityAPI.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    private static final Logger LOG = Log.logger(RestServer.class);
+
+    @GET
+    @Timed
+    @Produces(APPLICATION_JSON_WITH_CHARSET)
+    public String get(@Context GraphManager manager,
+                      @PathParam("graph") String graph,
+                      @QueryParam("v1") String v1,
+                      @QueryParam("v2") String v2,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/JaccardSimilarityAPI.java]: improve var name

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/SameNeighborsAPI.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                      @QueryParam("max_degree")
+                      @DefaultValue(DEFAULT_DEGREE) long degree,
+                      @QueryParam("limit")
+                      @DefaultValue(DEFAULT_PATHS_LIMIT) long limit) {
+        LOG.debug("Graph [{}] get same neighbors between '{}' and '{}' with " +
+                  "direction {}, edge label {}, max degree '{}' and limit '{}'",
+                  graph, v1, v2, direction, edgeLabel, degree, limit);
+
+        Id sourceId = VertexAPI.checkAndParseVertexId(v1);
+        Id targetId = VertexAPI.checkAndParseVertexId(v2);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/SameNeighborsAPI.java]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java [hunk 3]
```diff
@@ -171,6 +172,52 @@ public HugeGraph graph() {
...
+                                 String label, long degree, long limit) {
+        E.checkNotNull(v1, "vertex id");
+        E.checkNotNull(v2, "vertex id");
+        E.checkNotNull(direction, "direction");
+        checkDegree(degree);
+        checkLimit(limit);
+
+        Id labelId = this.getEdgeLabelId(label);
+
+        List<Id> sourceNeighbors = IteratorUtils.toList(this.adjacentVertices(
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java]: toSet

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java [hunk 4]
```diff
@@ -171,6 +172,52 @@ public HugeGraph graph() {
...
+                                   v1, direction, labelId, degree));
+        List<Id> targetNeighbors = IteratorUtils.toList(this.adjacentVertices(
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java [hunk 5]
```diff
@@ -171,6 +172,52 @@ public HugeGraph graph() {
...
+                                   v2, direction, labelId, degree));
+        Set<Id> sameNeighbors = new HashSet<>(CollectionUtils.intersection(
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java]: convert via "(Set<Id>)"

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SingleSourceShortestPathTraverser.java [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+
+        public ShortestPaths shortestPaths() {
+            return this.foundNodes;
+        }
+
+        private double edgeWeight(HugeEdge edge) {
+            double edgeWeight;
+            if (this.weight == null ||
+                !edge.property(this.weight).isPresent()) {
+                edgeWeight = 1;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SingleSourceShortestPathTraverser.java]: 1.0

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java [hunk 7]
```diff
@@ -171,6 +172,52 @@ public HugeGraph graph() {
...
+    public double jaccardSimilarity(Id v1, Id v2, Directions dir,
+                                    String label, long degree) {
+        E.checkNotNull(v1, "source vertex id");
+        E.checkNotNull(v2, "target vertex id");
+        E.checkNotNull(dir, "direction");
+        checkDegree(degree);
+
+        Id labelId = this.getEdgeLabelId(label);
+
+        List<Id> sourceNeighbors = IteratorUtils.toList(this.adjacentVertices(
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SingleSourceShortestPathTraverser.java [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+                         * 2. Already exist path for 'target' and current
+                         *    path is shorter, update path for 'target'
+                         */
+                        this.findingNodes.put(target, nw);
+                    }
+                }
+            }
+
+            List<NodeWithWeight> sorted = this.findingNodes.values().stream()
+                                          .sorted(Comparator.comparing
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SingleSourceShortestPathTraverser.java]: call method HugeTraverser.topN or CollectionUtil.sortByValue

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java [hunk 9]
```diff
@@ -171,6 +172,50 @@ public HugeGraph graph() {
...
+        E.checkNotNull(dir, "direction");
+        checkDegree(degree);
+
+        Id labelId = this.getEdgeLabelId(label);
+
+        Set<Id> sourceNeighbors = IteratorUtils.set(this.adjacentVertices(
+                                  vertex, dir, labelId, degree));
+        Set<Id> targetNeighbors = IteratorUtils.set(this.adjacentVertices(
+                                  other, dir, labelId, degree));
+        int interNum = CollectionUtils.intersection(sourceNeighbors,
```
[the remark this answers is not in the document]
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/HugeTraverser.java:211]: adjacentVertices() returns Iterator. Convert to list or set is necessary

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SingleSourceShortestPathTraverser.java [hunk 10]
```diff
@@ -76,6 +76,7 @@ public NodeWithWeight weightedShortestPath(Id sourceV, Id targetV,
                                                long skipDegree, long capacity) {
         E.checkNotNull(sourceV, "source vertex id");
         E.checkNotNull(dir, "direction");
+        E.checkNotNull(weight, "weight property");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SingleSourceShortestPathTraverser.java:79]: seems it's not checked at upper layer

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/WeightedShortestPathAPI.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+        LOG.debug("Graph [{}] get weighted shortest path between '{}' and " +
+                  "'{}' with direction {}, edge label {}, weight property {}, " +
+                  "max degree '{}', skip degree '{}', capacity '{}', " +
+                  "and with vertex '{}'",
+                  graph, source, target, direction, edgeLabel, weight, degree,
+                  skipDegree, capacity, withVertex);
+
+        Id sourceId = VertexAPI.checkAndParseVertexId(source);
+        Id targetId = VertexAPI.checkAndParseVertexId(target);
+        Directions dir = Directions.convert(EdgeAPI.parseDirection(direction));
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/WeightedShortestPathAPI.java:85]: check weight

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java [hunk 12]
```diff
@@ -245,4 +244,62 @@ private boolean finished() {
...
+                if (!this.vertices().get(i).equals(
+                     otherVertices.get(size - i - 1))) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    }
+
+    public static class PathSet extends HashSet<Path> {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/SubGraphTraverser.java:295]: prefer move to HugeTraverser.java

**Pull request merged**